docs/sat-solver.md

```diff
@@ -0,0 +1,383 @@
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{j} \in \{0,1\}, a_{ij},b_{i}, c_{j} \in N_{0}^{+}, i \in M">
+
+<img src="https://render.githubusercontent.com/render/math?math=M = 1, \dots, m">
+
+Where:
+
+- Each <img src="https://render.githubusercontent.com/render/math?math=c_{j}">
+  is a non-negative integer cost, associated with variable
+  <img src="https://render.githubusercontent.com/render/math?math=x_{j}, j \in N">.
+- <img src="https://render.githubusercontent.com/render/math?math=a_{ij}">
+  denotes the coefficients of the literals
+  <img src="https://render.githubusercontent.com/render/math?math=l_{j}">
+  in the set m of linear constraints (a literal being a Boolean variable).
+
+In our case, a package _x_ belongs to the repository _P_, where _P_ is the set
+of tuples
+<img src="https://render.githubusercontent.com/render/math?math=P = \{ (repoUrl, name, semverRange, ns), \dots \}">.
+
+## Constraints definitions
+
+We will define Pseudo-Boolean functions that will be optimized by selecting
+their input values. This Pseudo-Boolean functions can then be implemented in
+Hypper easily.
+
+### Installation
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{1}"> is installed:
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{1} \geq 1">,
+with <img src="https://render.githubusercontent.com/render/math?math=x_{1} \in P">
+
+For simplicity, the tuple of package
+<img src="https://render.githubusercontent.com/render/math?math=x = (repoUrl, name, semverRange = 1.0, ns)">
+is represented as
+<img src="https://render.githubusercontent.com/render/math?math=x_{1}">.
+
+
+### Depends
+
+<img src="https://render.githubusercontent.com/render/math?math=p_{1}"> depends on
+<img src="https://render.githubusercontent.com/render/math?math=x_{1}">:
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{1} - p_{1} \geq 0">
+
+Installing <img src="https://render.githubusercontent.com/render/math?math=p_{1}"> implies
+installing <img src="https://render.githubusercontent.com/render/math?math=x_{1}">.
+Although <img src="https://render.githubusercontent.com/render/math?math=x_{1}"> may be
+installed without <img src="https://render.githubusercontent.com/render/math?math=p_{1}">.
+
+
+### Depends on multiple versions
+
+If <img src="https://render.githubusercontent.com/render/math?math=p_{1}">
+depends on multiple versions of _x_, for example,
+<img src="https://render.githubusercontent.com/render/math?math=x_{1}"> and
+<img src="https://render.githubusercontent.com/render/math?math=x_{2}">:
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{1} %2B x_{2} - p_{1} \geq 0">
+
+
+### Conflicts
+
+If <img src="https://render.githubusercontent.com/render/math?math=y_{2}"> conflicts with
+<img src="https://render.githubusercontent.com/render/math?math=x_{1}">:
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{1} %2B y_{3} \leq 1">
+
+
+### Provides
+
```

Comment:
Not sure if a `Provides` needs to enter the formula.
If `y` requires `p` and `x` provides `p`, then it's reduced to `y` requires `x`

Reply:
Maybe the definition of Provides is not clear enough. The idea is to use it to create aliases (e.g: chart `foo` depends on chart `relational-db`, and then we have chart `mysql` that provides `relational-db`, and `postgresql` that provides `relational-db`).
The solver needs to be fed with the Provides relation, which is unknown to it, so it can find `x`, in this case `mysql` or `postgresql`.

Reply:
We need some interactive stuff for that in case there are multiple candidates either installed on the cluster (e.g. mysql and postgresql) which provide the same or if there is nothing installed on the cluster and there are multiple options available on the repositories that would satisfy the dependency.

Reply:
Interactive should be avoided imho, if other heuristics can be used to take a decision.

Reply:
@kkaempf agree, but in some cases you have packages that have the same "value" to the product (e.g. mariadb vs. mysql for wordpress etc.) in that case we have to interact with the user or be very opinionated. We should have a list on how to prioritise in such cases (e.g. if anything matches on the cluster use that [override option is needed here], otherwise use our preference)

Reply:
Also, we maybe should be more clear in the document that the relations "provides" and "conflicts" are not supported yet by Hypper charts, in the "package" definition at the beginning of the document. I was just listing them to help reason about the whole thing.

Reply:
@dragonchaser thinking about our target group (cluster admins), I agree. They should be able to decide mariadb vs. mysql.
But that doesn't hold for other target groups :wink:

Reply:
(clarified that "Provides, Conflicts" are yet not implemented in Hypper with https://github.com/rancher-sandbox/hypper/commit/3046654dd7d03da38bbfa24ce448e3b77cd5207a).